### ISSUE TITLE
🐛 [firebase_auth] signInWithCredential method does not emit authStateChanges event

### ISSUE THREAD

**Issue by aaort** [firebase/flutterfire#10159]
## Bug report

**Describe the bug**
This is issue is describing the same behavior as #3968, however, it was closed without any ways of resolving it provided.
I'm trying to authenticate users with phone number and [provider](https://pub.dev/packages/provider) package by globally listening to `authStateChanges` event stream to take appropriate steps if users is logged in or not.

The problems I'm having is that `authStateChanges` does not notify when new user logs in with credentials (using `signInWithCredential` method) but it works as expected with anonymous authentication.

And yes, I can tell for sure that user actually was able to log in because:
1. I can see the logs telling me about that.
2. After performing hot restart user is not null anymore and app does successfully navigate me to the home screen.

### Steps to reproduce

1. Create flutter project.
3. Add Firebase to your flutter project.
4. Enable phone number authentication inside firebase project console under 'sign-in method' tab.
5. Add Provider package to your flutter project (not actually required, could be any other approach for state management).
6. Setup StreamProvider which listens for `FirebaseAuth.instance.authStateChanges`.
7. And implement UI for authentication with phone number.

### Expected behavior

Expected behavior would be for  `authStateChanges` stream  to notify about new user login event as it does with anonymous sign in.

### Sample project

Unfortunately I can't share repo URL here because repo is private but I tried to minimize my code as much as possible to make it easy to work with.

<details><summary>main.dart</summary>

```dart
import 'package:firebase_core/firebase_core.dart';
import 'package:flutter/material.dart';
import 'package:to_buy/app_navigator.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  await Firebase.initializeApp();

  runApp(const App());
}

class App extends StatelessWidget {
  const App({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: AppNavigator(),
    );
  }
}

```
</details>

<details><summary>app_navigator.dart</summary>

```dart
import 'package:firebase_auth/firebase_auth.dart' show FirebaseAuth, User;
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:to_buy/screens/index.dart' show Home, PhoneNumberSignIn;

class AppNavigator extends StatelessWidget {
  const AppNavigator({super.key});

  @override
  Widget build(BuildContext context) {
    return StreamProvider<User?>.value(
      value: FirebaseAuth.instance.authStateChanges(),
      initialData: null,
      builder: (context, _) {
        if (context.watch<User?>() != null) {
          return const Home();
        } else {
          return const PhoneNumberSignIn();
        }
      },
    );
  }
}
```
</details>

<details><summary>phone_number_sign_in.dart</summary>

```dart
import 'package:firebase_auth/firebase_auth.dart' show FirebaseAuth;
import 'package:flutter/material.dart';
import 'package:to_buy/screens/index.dart' show Verification;

final _auth = FirebaseAuth.instance;

class PhoneNumberSignIn extends StatefulWidget {
  const PhoneNumberSignIn({super.key});

  @override
  State<PhoneNumberSignIn> createState() => _PhoneNumberSignInState();
}

class _PhoneNumberSignInState extends State<PhoneNumberSignIn> {
  String? _phoneNumber;

  void navigateToVerification(String verificationId) {
    Navigator.of(context).pushReplacement(
      MaterialPageRoute(
        builder: (context) => Verification(verificationId),
      ),
    );
  }

  _verifyPhoneNumber() {
    _auth.verifyPhoneNumber(
      phoneNumber: _phoneNumber!,
      verificationCompleted: (_) {},
      verificationFailed: (_) {},
      codeSent: (String verificationId, int? _) async {
        navigateToVerification(verificationId);
      },
      codeAutoRetrievalTimeout: (_) {},
    );
  }

  _onChanged(String value) => setState(() => _phoneNumber = value);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          TextField(
            onChanged: _onChanged,
            autofocus: true,
            keyboardType: TextInputType.phone,
          ),
          TextButton(
            onPressed: _verifyPhoneNumber,
            child: const Text('Sign in'),
          ),
        ],
      ),
    );
  }
}
```
</details>

<details><summary>verification.dart</summary>

```dart
import 'package:firebase_auth/firebase_auth.dart'
    show FirebaseAuth, PhoneAuthProvider;
import 'package:flutter/material.dart';

final _auth = FirebaseAuth.instance;

class Verification extends StatefulWidget {
  final String? verificationId;

  const Verification(this.verificationId, {super.key});

  @override
  State<Verification> createState() => _VerificationState();
}

class _VerificationState extends State<Verification> {
  _onCodeChanged(String code) {
    if (code.length == 6) onSignIn(code);
  }

  onSignIn(String code) async {
    if (widget.verificationId == null || code.length < 6) return;
    final credential = PhoneAuthProvider.credential(
      verificationId: widget.verificationId!,
      smsCode: code,
    );
  await _auth.signInWithCredential(credential);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: TextField(
          maxLength: 6,
          onChanged: _onCodeChanged,
          keyboardType: TextInputType.number,
          autofocus: true,
        ),
      ),
    );
  }
}

```
</details>

---

## Additional context

I was able to reproduce this behavior on Android Physical device (Xiaomi Mi A2) and iOS simulator (iPhone 14 Pro).

---

### Flutter doctor

<details><summary>Output of `flutter doctor`</summary>

```
[✓] Flutter (Channel stable, 3.3.10, on macOS 12.6.2 21G320 darwin-x64, locale en-AM)
    • Flutter version 3.3.10 on channel stable at /Users/Art/tools/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 135454af32 (6 days ago), 2022-12-15 07:36:55 -0800
    • Engine revision 3316dd8728
    • Dart version 2.18.6
    • DevTools version 2.15.0

[!] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/Art/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] VS Code (version 1.74.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.56.0

[✓] Connected device (4 available)
    • Mi A2 (mobile)         • 25d10fd                              • android-arm64  • Android 10 (API 29)
    • iPhone 14 Pro (mobile) • 2A2274B5-1518-4283-AB2B-D3C2BEE3E2CA • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-2 (simulator)
    • macOS (desktop)        • macos                                • darwin-x64     • macOS 12.6.2 21G320 darwin-x64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 108.0.5359.124

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```
</details>

---

### Flutter dependencies

<details><summary>Output of `flutter pub deps --style=compact`</summary>

```
Dart SDK 2.18.6
Flutter SDK 3.3.10
to_buy 1.0.0+1

dependencies:
- cloud_firestore 4.2.0 [cloud_firestore_platform_interface cloud_firestore_web collection firebase_core firebase_core_platform_interface flutter meta]
- cloud_functions 4.0.6 [cloud_functions_platform_interface cloud_functions_web firebase_core firebase_core_platform_interface flutter]
- cupertino_icons 1.0.5
- firebase_analytics 10.0.7 [firebase_analytics_platform_interface firebase_analytics_web firebase_core firebase_core_platform_interface flutter]
- firebase_auth 4.2.1 [firebase_auth_platform_interface firebase_auth_web firebase_core firebase_core_platform_interface flutter meta]
- firebase_core 2.4.0 [firebase_core_platform_interface firebase_core_web flutter meta]
- firebase_messaging 14.1.4 [firebase_core firebase_core_platform_interface firebase_messaging_platform_interface firebase_messaging_web flutter meta]
- flutter 0.0.0 [characters collection material_color_utilities meta vector_math sky_engine]
- keyboard_dismisser 3.0.0 [flutter]
- provider 6.0.5 [collection flutter nested]
- rxdart 0.27.7
- verification_code_field 0.0.1 [flutter]

dev dependencies:
- flutter_lints 2.0.1 [lints]
- flutter_test 0.0.0 [flutter test_api path fake_async clock stack_trace vector_math async boolean_selector characters collection matcher material_color_utilities meta source_span stream_channel string_scanner term_glyph]

transitive dependencies:
- _flutterfire_internals 1.0.10 [cloud_firestore_platform_interface cloud_firestore_web collection firebase_core firebase_core_platform_interface flutter meta]
- async 2.9.0 [collection meta]
- boolean_selector 2.1.0 [source_span string_scanner]
- characters 1.2.1
- clock 1.1.1
- cloud_firestore_platform_interface 5.9.2 [_flutterfire_internals collection firebase_core flutter meta plugin_platform_interface]
- cloud_firestore_web 3.1.1 [_flutterfire_internals cloud_firestore_platform_interface collection firebase_core firebase_core_web flutter flutter_web_plugins js]
- cloud_functions_platform_interface 5.1.25 [firebase_core flutter meta plugin_platform_interface]
- cloud_functions_web 4.3.14 [cloud_functions_platform_interface firebase_core firebase_core_web flutter flutter_web_plugins js]
- collection 1.16.0
- fake_async 1.3.1 [clock collection]
- firebase_analytics_platform_interface 3.3.15 [_flutterfire_internals firebase_core flutter meta plugin_platform_interface]
- firebase_analytics_web 0.5.1+6 [_flutterfire_internals firebase_analytics_platform_interface firebase_core firebase_core_web flutter flutter_web_plugins js]
- firebase_auth_platform_interface 6.11.5 [_flutterfire_internals collection firebase_core flutter meta plugin_platform_interface]
- firebase_auth_web 5.2.1 [firebase_auth_platform_interface firebase_core firebase_core_web flutter flutter_web_plugins http_parser intl js meta]
- firebase_core_platform_interface 4.5.2 [collection flutter flutter_test meta plugin_platform_interface]
- firebase_core_web 2.0.2 [firebase_core_platform_interface flutter flutter_web_plugins js meta]
- firebase_messaging_platform_interface 4.2.8 [_flutterfire_internals firebase_core flutter meta plugin_platform_interface]
- firebase_messaging_web 3.2.9 [_flutterfire_internals firebase_core firebase_core_web firebase_messaging_platform_interface flutter flutter_web_plugins js meta]
- flutter_web_plugins 0.0.0 [flutter js characters collection material_color_utilities meta vector_math]
- http_parser 4.0.2 [collection source_span string_scanner typed_data]
- intl 0.17.0 [clock path]
- js 0.6.4
- lints 2.0.1
- matcher 0.12.12 [stack_trace]
- material_color_utilities 0.1.5
- meta 1.8.0
- nested 1.0.0 [flutter]
- path 1.8.2
- plugin_platform_interface 2.1.3 [meta]
- sky_engine 0.0.99
- source_span 1.9.0 [collection path term_glyph]
- stack_trace 1.10.0 [path]
- stream_channel 2.1.0 [async]
- string_scanner 1.1.1 [source_span]
- term_glyph 1.2.1
- test_api 0.4.12 [async boolean_selector collection meta source_span stack_trace stream_channel string_scanner term_glyph matcher]
- typed_data 1.3.1 [collection]
- vector_math 2.1.2
```
</details>

**Comment by darshankawar** [comment]
Thanks for the report @aaort 
I tried to use the code sample you provided, but I was unable to properly replicate it.

<details>
<summary> complete code sample </summary>

```

import 'package:firebase_auth/firebase_auth.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:triage/firebase_options.dart';
//import 'package:to_buy/app_navigator.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await Firebase.initializeApp(options: DefaultFirebaseOptions.currentPlatform);
  runApp(const App());
}

class App extends StatelessWidget {
  const App();

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: AppNavigator(),
    );
  }
}

class AppNavigator extends StatelessWidget {
  const AppNavigator();

  @override
  Widget build(BuildContext context) {
    return StreamProvider<User?>.value(
      value: FirebaseAuth.instance.authStateChanges(),
      initialData: null,
      builder: (context, _) {
        if (context.watch<User?>() != null) {
          return Home();
        } else {
          return const PhoneNumberSignIn();
        }
      },
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    // TODO: implement build
    return Scaffold(
      body: Center(
        child: Text('Home'),
      )
    );
  }

}

final _auth = FirebaseAuth.instance;

class PhoneNumberSignIn extends StatefulWidget {
  const PhoneNumberSignIn();

  @override
  State<PhoneNumberSignIn> createState() => _PhoneNumberSignInState();
}

class _PhoneNumberSignInState extends State<PhoneNumberSignIn> {
  String? _phoneNumber;

  void navigateToVerification(String verificationId) {
    Navigator.of(context).pushReplacement(
      MaterialPageRoute(
        builder: (context) => Verification(verificationId),
      ),
    );
  }

  _verifyPhoneNumber() {
    _auth.verifyPhoneNumber(
      phoneNumber: _phoneNumber!,
      verificationCompleted: (_) {},
      verificationFailed: (_) {},
      codeSent: (String verificationId, int? _) async {
        navigateToVerification(verificationId);
      },
      codeAutoRetrievalTimeout: (_) {},
    );
  }

  _onChanged(String value) => setState(() => _phoneNumber = value);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          TextField(
            onChanged: _onChanged,
            autofocus: true,
            keyboardType: TextInputType.phone,
          ),
          TextButton(
            onPressed: _verifyPhoneNumber,
            child: const Text('Sign in'),
          ),
        ],
      ),
    );
  }
}

class Verification extends StatefulWidget {
  final String? verificationId;

  const Verification(this.verificationId);

  @override
  State<Verification> createState() => _VerificationState();
}

class _VerificationState extends State<Verification> {
  _onCodeChanged(String code) {
    if (code.length == 6) onSignIn(code);
  }

  onSignIn(String code) async {
    if (widget.verificationId == null || code.length < 6) return;
    final credential = PhoneAuthProvider.credential(
      verificationId: widget.verificationId!,
      smsCode: code,
    );
    await _auth.signInWithCredential(credential);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: TextField(
          maxLength: 6,
          onChanged: _onCodeChanged,
          keyboardType: TextInputType.number,
          autofocus: true,
        ),
      ),
    );
  }
}

```

</details>

Can you try the [plugin example](https://github.com/firebase/flutterfire/tree/master/packages/firebase_auth/firebase_auth/example) and see if using it, you get same behavior or not ?

**Comment by aaort** [comment]
Sure, Should I test it with already initialized firebase project or replace it with mine ?
I've already tried to verify phone number with default firebase project but getting SHA-1 related error. I don't know how to access firebase project settings used in the example app...

**Comment by darshankawar** [comment]
> Should I test it with already initialized firebase project or replace it with mine ?

You may try with your `firebase_options.dart` file and no need to change any other setup I think.

**Comment by aaort** [comment]
I tried, and it worked...

I added new android app in firebase console, downloaded `google-services.json` file and replace old file with it didn't event touch `firebase_options.dart` file.

After phone number verification `authStateChanges` listener receives new event along with newly logged in user and I can see user in firebase console.

So there is obviously something wrong on my side, gonna close this and try to figure out the problem.
Thank you very much for your help.
